docs/kbs_attestation_protocol.md

```diff
@@ -1,220 +1,412 @@
-The fresh number passed to KBC. KBC needs to place it in the evidence sent to KBS in the next
-step to prevent replay attacks.
+- `tee-pubkey`
 
-- `extra-params`
+After KBC receives the attestation challenge, an ephemeral asymmetric key pair
+is generated in HW-TEE. The private key is stored in HW-TEE. The public key and
+its description information are exported and placed in the `tee-pubkey` field
+and sent to the KBS together with the attestation evidence. The hash of the
+`tee-pubkey` field must be included in the custom field of HW-TEE evidence and
+signed by HW-TEE hardware. This public key is valid until the next time the KBC
+receives an attestation challenge from the KBS.
 
-The reserved extra parameter field which is used to pass the additional information provided by
-the KBS when some specific HW-TEE needs to be attested.
+The `tee-pubkey` follows the  [JSON Web Key](https://www.rfc-editor.org/rfc/rfc7517)
+format, as described in the [key format](#key-format) section.
 
-## Attestation
+- `tee-evidence`
 
-After receiving the invitation challenge, KBC collects the attestation evidence from the HW-TEE
-platform and organizes it into the following payload format:
+The attestation evidence generated by the HW-TEE platform software and hardware
+in the KBC's execution environment.
+The `tee-evidence` formats depend on the TEE and are typically defined by the
+Attestation-Service.
 
-```json
+The KBS does not parse or analyze the attestation evidence, it forwards it to
+the Attestation-Service for verification.
+
+## `Response`
+
+Upon succcessful attestation, the KBC can request resources from the KBS, by
+sending HTTP `GET` requests to it.
+
+If the KBS approves the request, it responds to the KBC by sending a `Response`
+payload that follows the [JSON Web Encryption](https://www.rfc-editor.org/rfc/rfc7516)
+format:
+
+``` json
 {
-    "evidence": {
-        /* The nonce in the Challenge; its hash needs to be included in HW-TEE evidence and signed by HW-TEE hardware. */
-        "nonce": "",
-        /* TEE type name */
-        "tee": "",
-        /* The public key generated by KBC in HW-TEE, it is valid until the next time an attestation is required. Its hash needs to be included in HW-TEE evidence and signed by HW-TEE hardware. */
-        "tee-pubkey": {
-            "algorithm": "",
-            "pubkey-length": "",
-            "pubkey": ""
-        },
-        /* The content of evidence. Its format is specified by Attestation-Service. */
-        "tee-evidence": {}
-    }
+    /* JOSE Header */
+    "alg": "$alg",
+    "enc": "$enc"
+}.
+{
+    /*
+     * The output of the KBS service API.
+     * It must be encrypted with the ephemeral symmetric key generated by the KBS.
+     */
+    "ciphertext": "$response_output",
+    "iv": "$initialization_vector",
+    /*
+     * The symmetric key used to encrypt `ciphertext`, which is encrypted by
+     * the HW-TEE's public key.
+     */
+    "encrypted_key": "$enc_kbs_symkey"
 }
 ```
```

Comment:
JSON lint tells me this isn't valid JSON. Is this intended to be an array? Can the JOSE header object be a field within the root object?

Reply:
@slp Yes, the example is incorrectly formatted and the intention is for it to be an array, yes. So it should really look like that:

```json
[{
	"alg": "$alg",
	"enc": "$enc"
}, {
	"ciphertext": "$response_output",
	"iv": "$initialization_vector",
	"encrypted_key": "$enc_kbs_symkey"
}]
```

Reply:
I'm not familiar with the JOSE specs. Is it required to be found in this form? As ordering in an array isn't guaranteed, consuming this payload would require to peek into the object to find out its format.

Would it be possible for the JOSE header to be an object in named field? Something like:

```json
{
	"ciphertext": "$response_output",
	"iv": "$initialization_vector",
	"encrypted_key": "$enc_kbs_symkey",
	"jose_params": {
		"alg": "$alg",
		"enc": "$enc"
	}
}
```

Reply:
> I'm not familiar with the JOSE specs. Is it required to be found in this form? As ordering in an array isn't guaranteed, consuming this payload would require to peek into the object to find out its format.
> 
> Would it be possible for the JOSE header to be an object in named field? Something like:
> 
> ```json
> {
> 	"ciphertext": "$response_output",
> 	"iv": "$initialization_vector",
> 	"encrypted_key": "$enc_kbs_symkey",
> 	"jose_params": {
> 		"alg": "$alg",
> 		"enc": "$enc"
> 	}
> }
> ```

AFAIK, the architecture of a JOSE object should look like
```rust
BASE64URL_ENCODE({"alg": "$alg","enc": "$enc"}).
BASE64URL_ENCODE($enc_kbs_symkey).
BASE64URL_ENCODE($initialization_vector).
BASE64URL_ENCODE($response_output).
BASE64URL_ENCODE($authentication_tag)
```

Here `$authentication_tag` is extra result of specific symmetric encryption (s.t. AEAD) besides cipher text. Algorithms like `AES/GCM` will produce such authentication data. And care about each adjacent two base64url-encoded part are connected by a dot `.`

I think the json format shown in the document may be for understanding the fields of the JWE used here. Am I right? @sameo 

Reference to [JOSE Cookbook](https://datatracker.ietf.org/doc/html/rfc7520#section-5.1.5) and [JWE Spec](https://www.rfc-editor.org/rfc/rfc7516#section-2).

Reply:
Sorry for the delayed answer, post LPC has been busy for me...

@Xynnn007 you're partly right, although what you're describing is the [compact JWE serialization format](https://www.rfc-editor.org/rfc/rfc7516#section-7.1).
But I think we want to use the [flattened JWE JSON serialization format](https://www.rfc-editor.org/rfc/rfc7516#section-7.2.2). 

With that format, our `Response` JSON-formatted payload would then look like that:

```json
{
    "protected": "$jose_header",
    "encrypted_key": "$encrypted_key",
    "iv": "$iv",
    "ciphertext": "$ciphertext",
    "tag": "$tag"
}
```

Where:

```shell
jose_header = BASE64URL_ENCODE({"alg": "$alg","enc": "$enc"})
encrypted_key = BASE64URL_ENCODE($enc_kbs_symkey)
iv = BASE64URL_ENCODE($initialization_vector)
ciphertext = BASE64URL_ENCODE($response_output)

# tag is optional and depend on the encryption algorithm.
tag = BASE64URL_ENCODE($authentication_tag) 
```

I find it a little convoluted, but this really is what the spec defines. FWIW there are some on going, recent work on a rust crate for JWE and JWK: https://github.com/RustCrypto/JOSE

If we're fine with that definition, I'll update that PR accordingly.

Reply:
Thanks @slp. I updated the document and the openAPI description as well. I think we *may* be good now :) PTAL.